File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates, target):
        def sid(i,n,li,lis,target,res):
            if target==0:
                res.append(list(lis))
                return
            for j in range(i,n):
                if j>i and li[j]==li[j-1]:
                    continue
                if li[j]>target:
                    break
                lis.append(li[j])
                sid(j+1,n,li,lis,target-li[j],res)
                lis.pop()
        candidates.sort()
        ans = []
        ds = []
        sid(0, len(candidates), candidates,ds,target, ans)
        return ans

        

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
    public: 
    void sum(int ind, vector<int> &v, vector<int> &ve,vector<vector<int>> &res,int target){
    if(target==0){
        res.push_back(ve);
        return;
    }
    for(int i=ind;i<v.size();i++){
        if(i>ind && v[i]==v[i-1]){
            continue;
        }
        if(v[i]>target){
            break;
        }
        ve.push_back(v[i]);
        sum(i+1,v,ve,res,target-v[i]);
        ve.pop_back();
    }
}
    public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans; 
        vector<int> ds; 
        sum(0, candidates, ds, ans, target); 
        return ans; 
    }
};